File: system/backend/app/services/pluralkit.py
"""
PluralKit API integration service
Handles all interactions with the PluralKit API
"""

import httpx
from typing import List, Dict, Any
from app.core.config import PLURALKIT_BASE_URL, get_pluralkit_headers, CACHE_TTL
from app.utils.cache import get_from_cache, set_in_cache

# Special member display names
SPECIAL_DISPLAY_NAMES = {
    "answer": "Answer Machine",
    "system": "Unsure",
    "sleeping": "I am sleeping"
}

async def get_system() -> Dict[str, Any]:
    """Get system information from PluralKit"""
    cache_key = "system"
    if (cached := get_from_cache(cache_key)):
        return cached
    
    async with httpx.AsyncClient() as client:
        resp = await client.get(
            f"{PLURALKIT_BASE_URL}/systems/@me",
            headers=get_pluralkit_headers()
        )
        resp.raise_for_status()
        data = resp.json()
        set_in_cache(cache_key, data, CACHE_TTL)
        return data

async def get_members() -> List[Dict[str, Any]]:
    """Get all system members from PluralKit"""
    cache_key = "members"
    if (cached := get_from_cache(cache_key)):
        return cached
    
    # Get raw members from API
    base_cache_key = "members_raw"
    if not (cached_raw := get_from_cache(base_cache_key)):
        async with httpx.AsyncClient() as client:
            resp = await client.get(
                f"{PLURALKIT_BASE_URL}/systems/@me/members",
                headers=get_pluralkit_headers()
            )
            resp.raise_for_status()
            cached_raw = resp.json()
            set_in_cache(base_cache_key, cached_raw, CACHE_TTL)
    
    # Process special members
    processed_members = []
    for member in cached_raw:
        member_name = member.get("name")
        
        if member_name in SPECIAL_DISPLAY_NAMES:
            special_member = {
                **member,
                "display_name": SPECIAL_DISPLAY_NAMES[member_name],
                "is_special": True,
                "original_name": member_name
            }
            processed_members.append(special_member)
        else:
            processed_members.append(member)
    
    set_in_cache(cache_key, processed_members, CACHE_TTL)
    return processed_members

async def get_fronters() -> Dict[str, Any]:
    """Get current fronters from PluralKit"""
    cache_key = "fronters"
    if (cached := get_from_cache(cache_key)):
        return cached
    
    async with httpx.AsyncClient() as client:
        resp = await client.get(
            f"{PLURALKIT_BASE_URL}/systems/@me/fronters",
            headers=get_pluralkit_headers()
        )
        resp.raise_for_status()
        data = resp.json()
        
        # Process special members in fronters
        if "members" in data:
            all_members = await get_members()
            
            processed_fronters = []
            for member in data["members"]:
                # Find processed member data
                processed_member = next(
                    (m for m in all_members if m.get("id") == member.get("id")),
                    member
                )
                processed_fronters.append(processed_member)
            
            data["members"] = processed_fronters
        
        set_in_cache(cache_key, data, CACHE_TTL)
        return data

async def set_front(member_ids: List[str]) -> Any:
    """
    Set the current front to the provided list of member IDs
    
    Args:
        member_ids: List of member IDs to set as fronting
    
    Returns:
        Response from PluralKit API (or None)
    """
    # Clear fronters cache
    set_in_cache("fronters", None, 0)
    
    async with httpx.AsyncClient() as client:
        resp = await client.post(
            f"{PLURALKIT_BASE_URL}/systems/@me/switches",
            headers=get_pluralkit_headers(),
            json={"members": member_ids}
        )
        
        if resp.status_code not in (200, 204):
            raise Exception(f"Failed to set front: {resp.status_code} - {resp.text}")
        
        return resp.json() if resp.content else None

async def get_switches(limit: int = 1000) -> List[Dict[str, Any]]:
    """Get recent switches from PluralKit"""
    cache_key = f"switches_{limit}"
    if (cached := get_from_cache(cache_key)):
        return cached
    
    async with httpx.AsyncClient() as client:
        resp = await client.get(
            f"{PLURALKIT_BASE_URL}/systems/@me/switches?limit={limit}",
            headers=get_pluralkit_headers()
        )
        resp.raise_for_status()
        data = resp.json()
        set_in_cache(cache_key, data, CACHE_TTL)
        return data

File: co/system/backend/src/services/pluralkit.ts
/**
 * Copyright (c) 2026 Clove Twilight
 * Licensed under the ESAL-2.0 Licence.
 * See LICENCE.md in the project root for full licence information.
 */

/**
 * PluralKit API integration service.
 * Handles all interactions with the PluralKit API.
 */

import axios from 'axios';

import { PLURALKIT_BASE_URL, getPluralkitHeaders, CACHE_TTL } from '../core/config.js';
import { getFromCache, setInCache } from '../utils/cache.js';

// Generic PluralKit JSON object — the upstream schema is large and not
// fully modeled here, mirroring the Python service's use of plain dicts.
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type PKObject = Record<string, any>;

// Special member display names
const SPECIAL_DISPLAY_NAMES: Record<string, string> = {
  answer: 'Answer Machine',
  system: 'Unsure',
  sleeping: 'I am sleeping',
};

/** Get system information from PluralKit */
export async function getSystem(): Promise<PKObject> {
  const cacheKey = 'system';
  const cached = getFromCache<PKObject>(cacheKey);
  if (cached) {
    return cached;
  }

  const { data } = await axios.get<PKObject>(`${PLURALKIT_BASE_URL}/systems/@me`, {
    headers: getPluralkitHeaders(),
  });
  setInCache(cacheKey, data, CACHE_TTL);
  return data;
}

/** Get all system members from PluralKit */
export async function getMembers(): Promise<PKObject[]> {
  const cacheKey = 'members';
  const cached = getFromCache<PKObject[]>(cacheKey);
  if (cached) {
    return cached;
  }

  // Get raw members from API
  const baseCacheKey = 'members_raw';
  let cachedRaw = getFromCache<PKObject[]>(baseCacheKey);
  if (!cachedRaw) {
    const { data } = await axios.get<PKObject[]>(`${PLURALKIT_BASE_URL}/systems/@me/members`, {
      headers: getPluralkitHeaders(),
    });
    cachedRaw = data;
    setInCache(baseCacheKey, cachedRaw, CACHE_TTL);
  }

  // Process special members
  const processedMembers: PKObject[] = cachedRaw.map((member) => {
    const memberName = member.name;

    if (memberName in SPECIAL_DISPLAY_NAMES) {
      return {
        ...member,
        display_name: SPECIAL_DISPLAY_NAMES[memberName],
        is_special: true,
        original_name: memberName,
      };
    }
    return member;
  });

  setInCache(cacheKey, processedMembers, CACHE_TTL);
  return processedMembers;
}

/** Get current fronters from PluralKit */
export async function getFronters(): Promise<PKObject> {
  const cacheKey = 'fronters';
  const cached = getFromCache<PKObject>(cacheKey);
  if (cached) {
    return cached;
  }

  const { data } = await axios.get<PKObject>(`${PLURALKIT_BASE_URL}/systems/@me/fronters`, {
    headers: getPluralkitHeaders(),
  });

  // Process special members in fronters
  if ('members' in data) {
    const allMembers = await getMembers();

    const processedFronters = (data.members as PKObject[]).map((member) => {
      return allMembers.find((m) => m.id === member.id) ?? member;
    });

    data.members = processedFronters;
  }

  setInCache(cacheKey, data, CACHE_TTL);
  return data;
}

/**
 * Set the current front to the provided list of member IDs.
 *
 * @param memberIds - List of member IDs to set as fronting
 * @returns Response body from the PluralKit API (or null if empty)
 * @throws Error if the PluralKit API call fails
 */
export async function setFront(memberIds: string[]): Promise<unknown> {
  // Clear fronters cache
  setInCache('fronters', null, 0);

  const resp = await axios.post(
    `${PLURALKIT_BASE_URL}/systems/@me/switches`,
    { members: memberIds },
    {
      headers: getPluralkitHeaders(),
      validateStatus: () => true, // handle non-2xx ourselves, like resp.raise_for_status() didn't run here
    },
  );

  if (resp.status !== 200 && resp.status !== 204) {
    throw new Error(`Failed to set front: ${resp.status} - ${JSON.stringify(resp.data)}`);
  }

  return resp.data ?? null;
}

/** Get recent switches from PluralKit */
export async function getSwitches(limit = 1000): Promise<PKObject[]> {
  const cacheKey = `switches_${limit}`;
  const cached = getFromCache<PKObject[]>(cacheKey);
  if (cached) {
    return cached;
  }

  const { data } = await axios.get<PKObject[]>(`${PLURALKIT_BASE_URL}/systems/@me/switches`, {
    headers: getPluralkitHeaders(),
    params: { limit },
  });
  setInCache(cacheKey, data, CACHE_TTL);
  return data;
}